When we do differential analysis in a Rna-seq expreiments we often do it in case vs control setting. We also know the sources of variation between these sample types. These sources of variations can be a drug treatment, certain phenotypic difference (cancerous vs healthy) etc. But on most of occasions we have some known and expected sources or variations such as experiments permormed on different batches, temparature of the day a particular experiments was performed etc. These kinds of variations are not biological rather technical in nature, and may impact the results of an Rna-seq experiment.  

So while dealing with the technical reasons, basically two possible scenario can occur; 1) one in which we know the sources of variations. 2) Secondly, we have only the count tables corresponding to case and control and the sources variation explain only a small proportions of the differnces between case and control. Package `DESeq2` has the ability to address above two situations.

We will discuss above two scenario. for that we obtain a count matrix and column meta data of 6 healthy heart and 6 failing heart samples from package `compGenomRData`.

```r
counts_file <- system.file('extdata/rna-seq/SRP021193.raw_counts.tsv',
                           package = 'compGenomRData')
colData_file <- system.file('extdata/rna-seq/SRP021193.colData.tsv',
                            package = 'compGenomRData')
                            
counts <- read.table(counts_file)
colData <- read.table(colData_file, header = T, sep = '\t',
                                  stringsAsFactors = TRUE)
```

We can have a look at how the samples cluster together(motivation again, is to check if the similar samples cluster together). For clustering first we need to normalize the counts matrix and we will use TPM as normalization measure. We would take the top 100 genes based on variation.

```r
geneLengths <- counts$width

rpk <- apply(subset(counts, select = c(-width)), 2,
              function(x) x/(geneLengths/1000))

tpm <- apply(rpk, 2, function(x) x / sum(as.numeric(x)) * 10^6)

selectedGenes <- names(sort(apply(tpm, 1, var),
                            decreasing = T)[1:100])
pheatmap(tpm[selectedGenes,],
         scale = 'row',
         annotation_col = colData,
         show_rownames = FALSE)
```
![image16](https://user-images.githubusercontent.com/85447250/218323957-a571ae41-e212-4191-9a9d-44ab71def482.png)

Fig. Heatmap showing the batch/technical effect on the samples.

In the figure above we can see that the case and control samples are not clearly separated. we have a scenario where the samples are clusterd based on the library selection variable rather than diagnosis and group. This is one such example of misclustering due to technical variations. `DESeq2` has the ability to address this factors and take in account the technical/batch effect on experiments and it will ineteresting to see the effect of proper design matrix on clustereing. 

```r
# get the raw counts
countData <- as.matrix(subset(counts, select = c(-width)))

# we account for the LibrarySelection and group in design matrix
dds <- DESeqDataSetFromMatrix(countData = countData,
                              colData = colData,
                              design = ~ LibrarySelection + group)
# running DESeq2
dds <- DESeq(dds)

# extracting results from DESeqDataSet object
DEresults <- results(dds, contrast = c('group', 'CASE', 'CTRL'))

# removing all the genes with NA value in padj column 
DE <- DEresults[!is.na(DEresults$padj),]

# Selecting all genes for which FDR < 10%
DE <- DE[DE$padj < 0.1,]

# selecting all genes DE$log2FoldChange > 1; it gives in genes which are either clearly upregulated or downregulated 
DE <- DE[abs(DE$log2FoldChange) > 1,]

# Selecting 100 such genes (after filtering from above parameters)
selectedGenes <- head(rownames(DE),100)

pheatmap(tpm[selectedGenes,],
         scale = 'row',
         annotation_col = colData,
         show_rownames = FALSE)
```

![image17](https://user-images.githubusercontent.com/85447250/218807711-6b05a122-febb-43a1-a0e4-ad69cff01dea.png)

Fig. clustering after including LibrarySelection in the design matrix to take account of technical reasons affecting the experiment.

From the above figure we can clearly see that once we account for the various technical reasons in the design matrix we get a clear separation of `CASE` and `CTRL` samples. Additionally, we can see that samples are clustered according the diagnosis as well which we expect. 







