### 

When we do differential analysis in a Rna-seq expreiments we often do it in case vs control setting. We also know the sources of variation between these sample types. These sources of variations can be a drug treatment, certain phenotypic difference (cancerous vs healthy) etc. But on most of occasions we have some known and expected sources or variations such as experiments permormed on different batches, temparature of the day a particular experiments was performed etc. These kinds of variations are not biological rather technical in nature, and may impact the results of an Rna-seq experiment.  

So while dealing with the technical reasons, basically two possible scenario can occur; 1) one in which we know the sources of variations. 2) Secondly, we have only the count tables corresponding to case and control and the sources variation explain only a small proportions of the differnces between case and control. Package `DESeq2` has the ability to address above two situations.

We will discuss above two scenario. for that we obtain a count matrix and column meta data of 6 healthy heart and 6 failing heart samples from package `compGenomRData`.

### When we have a known source of variation

```r
counts_file <- system.file('extdata/rna-seq/SRP021193.raw_counts.tsv',
                           package = 'compGenomRData')
colData_file <- system.file('extdata/rna-seq/SRP021193.colData.tsv',
                            package = 'compGenomRData')
                            
counts <- read.table(counts_file)
colData <- read.table(colData_file, header = T, sep = '\t',
                                  stringsAsFactors = TRUE)
```

We can have a look at how the samples cluster together(motivation again, is to check if the similar samples cluster together). For clustering first we need to normalize the counts matrix and we will use TPM as normalization measure. We would take the top 100 genes based on variation.

```r
geneLengths <- counts$width

rpk <- apply(subset(counts, select = c(-width)), 2,
              function(x) x/(geneLengths/1000))

tpm <- apply(rpk, 2, function(x) x / sum(as.numeric(x)) * 10^6)

selectedGenes <- names(sort(apply(tpm, 1, var),
                            decreasing = T)[1:100])
pheatmap(tpm[selectedGenes,],
         scale = 'row',
         annotation_col = colData,
         show_rownames = FALSE)
```
![image16](https://user-images.githubusercontent.com/85447250/218323957-a571ae41-e212-4191-9a9d-44ab71def482.png)

Fig. Heatmap showing the batch/technical effect on the samples.

In the figure above we can see that the case and control samples are not clearly separated. we have a scenario where the samples are clusterd based on the library selection variable rather than diagnosis and group. Ideally, a case sample should be a part of the contiguous `CASE` cluster while a control sample should be a part of contiguous `CTRL` cluster. This is one such example of misclustering due to technical variations. `DESeq2` has the ability to address this factors and take in account the technical/batch effect on experiments, and it will ineteresting to see the effect of proper design matrix on clustereing. 

```r
# get the raw counts
countData <- as.matrix(subset(counts, select = c(-width)))

# we account for the LibrarySelection and group in design matrix
dds <- DESeqDataSetFromMatrix(countData = countData,
                              colData = colData,
                              design = ~ LibrarySelection + group)
# running DESeq2
dds <- DESeq(dds)

# extracting results from DESeqDataSet object
DEresults <- results(dds, contrast = c('group', 'CASE', 'CTRL'))

# removing all the genes with NA value in padj column 
DE <- DEresults[!is.na(DEresults$padj),]

# Selecting all genes for which FDR < 10%
DE <- DE[DE$padj < 0.1,]

# selecting all genes DE$log2FoldChange > 1; it gives in genes which are either clearly upregulated or downregulated 
DE <- DE[abs(DE$log2FoldChange) > 1,]

# Selecting 100 such genes (after filtering from above parameters)
selectedGenes <- head(rownames(DE),100)

pheatmap(tpm[selectedGenes,],
         scale = 'row',
         annotation_col = colData,
         show_rownames = FALSE)
```

![image17](https://user-images.githubusercontent.com/85447250/218807711-6b05a122-febb-43a1-a0e4-ad69cff01dea.png)

Fig. clustering after including LibrarySelection in the design matrix to take account of technical reasons affecting the experiment.

From the above figure we can clearly see that once we account for the various technical reasons in the design matrix we get a clear separation of `CASE` and `CTRL` samples. Additionally, we can see that samples are clustered according the diagnosis as well which we expect. 

### When we don't know the source of variation, just the count matrix
 Cases when the potential source of variation is not known i.e., when the amount of variation is not explained by the known source of variation, and we left with just the count matrix to make a good guess of potential source of variations. In this scenario, package `RUVSeq` can help us to identify the source and variations and do the necessary normalization of the count table.Later on, we can implement these variations in the design formula  of `DESEq2`. But, first we need to get the data set.
 We will begin with  a lung carcinoma study, where a transcription factor (Ets homologous factor - EHF) is overexpressed and compared to the control samples with baseline EHF expression. We will take 5 case samples and 5 control samples. 
 
 ```r
 counts_file <- system.file('extdata/rna-seq/SRP049988.raw_counts.tsv',
                           package = 'compGenomRData')
colData_file <- system.file('extdata/rna-seq/SRP049988.colData.tsv',
                            package = 'compGenomRData')
counts <- read.table(counts_file)
colData <- read.table(colData_file, header = T,
                      sep = '\t', stringsAsFactors = TRUE)
colData$source_name <- ifelse(colData$group == 'CASE',
                              'EHF_overexpression', 'EHF_baseline')

geneLengths <- counts$width
rpk <- apply( subset(counts, select = c(-width)), 2,
              function(x) x/(geneLengths/1000))
tpm <- apply(rpk, 2, function(x) x / sum(as.numeric(x)) * 10^6)

selectedGenes <- names(sort(apply(tpm, 1, var),
                            decreasing = T)[1:100])
pheatmap(tpm[selectedGenes,],
         scale = 'row',
         annotation_col = colData,
         cutree_cols = 2,
         show_rownames = FALSE)
         
 > colData
              source_name group
CASE_1 EHF_overexpression  CASE
CASE_2 EHF_overexpression  CASE
CASE_3 EHF_overexpression  CASE
CASE_4 EHF_overexpression  CASE
CASE_5 EHF_overexpression  CASE
CTRL_1       EHF_baseline  CTRL
CTRL_2       EHF_baseline  CTRL
CTRL_3       EHF_baseline  CTRL
CTRL_4       EHF_baseline  CTRL
CTRL_5       EHF_baseline  CTRL        
 ```

![image18](https://user-images.githubusercontent.com/85447250/219115573-6f287502-214d-4af2-9784-910c9387c219.png)

Fig. cluster of samples based on EHF expression.

We can see that one of the case sample `case 5` is clustered with control group. This may be due to the some technical reason.
We will use `RUVSeq` to find the potential source of variations. If we look at the coldata (colData) object, we don't see any reason mentioned there. So, it may be due to batch effect, or some other technical reason. The point here is that we need to account for this source of variation in our count data and in design experiment, so that when we do DE analysis we get results conforming the paradigm of a RNAseq analysis.

```r
countData <- as.matrix(subset(counts, select = c(-width)))


## create a seqExpressionSet object using EDASeq package
set <- newSeqExpressionSet(counts = countData,
                           phenoData = colData)
```

Now let us first plot the RLE data for the raw counts

```r
par(mfrow = c(1,2))
plotRLE(set, outline=FALSE, ylim=c(-4, 4), col=as.numeric(colData$group))
plotPCA(set, col = as.numeric(colData$group), adj = 0.5, 
        ylim = c(-0.7, 0.5), xlim = c(-0.5, 0.5))

```
![image19](https://user-images.githubusercontent.com/85447250/219152881-c136fa1c-129c-4bba-8d27-84012012c521.png)

Fig. RLE and PCA plot of raw count matrix


Now plotting with the normalized count matrix (tpm)
```r
par(mfrow = c(1,2))
plotRLE(tpm, outline=FALSE, ylim=c(-4, 4), col=as.numeric(colData$group))
plotPCA(tpm, col=as.numeric(colData$group), adj = 0.5, 
        ylim = c(-0.3, 1), xlim = c(-0.5, 0.5))

```

![image20](https://user-images.githubusercontent.com/85447250/219154514-494aba1b-a44d-4375-99f8-b3de39730742.png)

Fig. RLE and PCA of normalized count matrix (tpm)

We can see from the figure above that the normalization has resulted in a much better distribution of gene counts across all samples (see RLE). Still, we see that the `case 5` is clustered along with control samples. We can use few different approaches using `RUVSeq` to point out the potential source of variation.

+ USing RUVg

One way of removing unwanted variation from the data sets is to make use of a set of genes whose expression does not change due to technical reasons, or atleast they are not expected to change because of technical reasons. One strategy is to add spike-In at the begining of experiments. Since, such spike-Ins are added in equal quantities in all samples, any significant deviation from the spike in expression level can be detected and corrected for. But, many experiments don't have spike in data available. In such cases, an empirical set of genes can be collected from the expression data by doing a differential expression analysis and discovering genes that are unchanged in the given conditions. These unchanged genes are used to clean up the data from systematic shifts in expression due to the unwanted sources of variation     











