
### Differential gene expression analysis 

Differential gene analysis compares a gene to thousands of other genes with the null hypothesis that the expression level of gene is same in two different samples. 

```r

library(DEseq2)
# to do a differential expression analysis using 'DESeq2' we need $3$ input objects, the count matrix, metadata of count 
# matrix, and a design formula

countData <- as.matrix(subset(counts, select = c(-width)))

colData <- read.table(coldata_file, header = TRUE, sep = "\t",
                      stringsAsFactors = TRUE) 
designformula <- "~ group"

# Creating a DEseqDataSet object 
dds <- DESeqDataSetFromMatrix(countData, colData = colData, 
                              design = as.formula(designformula) )

# we are taking out all such genes which does not have even have total sum of 1 across all samples (10 samples)

dds <- dds[rowSums(DESeq2::counts(dds)) > 1,]

# Now we can use DEseq() function from DESeq2 which will normalize the counts, estimate the dispersion values, and 
# compute the generalized linear model with the provided design formula. It does return an updated dds object.

dds <- DESeq(dds)
```

Now we can contrast and compare the sample based upon some variable. For instance, we have here "group" variable which describe
the `case` or `control` label for the samples. 

```r
DEresults = results(dds, contrast = c("group", "CASE", "CTRL"))

# arranging the result with decreasing order of p.values

DEresults <- DEresults[order(DEresults$pvalue),]
```

Above we have obtained a table containing the differential expression status of `case` samples compared to the `control` samples. It is important to note that the sequence of the elements provided in the contrast argument determines which group of samples are to be used as the `control`. This impacts the way the results are interpreted, for instance, if a gene is found up-
regulated (has a positive log2 fold change), the up-regulation status is only relative to the factor that is provided as `control`. In this case, we used samples from the “CTRL” group as `control` and contrasted the samples from the “CASE” group with respect to the “CTRL” samples. Thus genes with a positive log2 fold change are called up-regulated in the case samples with respect to the `control`, while genes with a negative log2 fold change are down-regulated in the `case` samples. Whether the deregulation is signiﬁcant or not, warrants assessment of the adjusted p-values. We can have a look at DEresult below.

```r
> DEresults
log2 fold change (MLE): group CASE vs CTRL 
Wald test p-value: group CASE vs CTRL 
DataFrame with 19097 rows and 6 columns
            baseMean log2FoldChange     lfcSE       stat       pvalue         padj
           <numeric>      <numeric> <numeric>  <numeric>    <numeric>    <numeric>
CYP2E1       4829889        9.36024  0.215223    43.4909  0.00000e+00  0.00000e+00
FCGBP       10349993       -7.57579  0.186433   -40.6355  0.00000e+00  0.00000e+00
ASGR2         426422        8.01830  0.216207    37.0863 4.67898e-301 2.87741e-297
GCKR          100183        7.82841  0.233376    33.5442 1.09479e-246 5.04945e-243
APOA5         438054       10.20248  0.312503    32.6477 8.64906e-234 3.19133e-230
...              ...            ...       ...        ...          ...          ...
CCDC195      20.4981      -0.215607   2.89255 -0.0745386           NA           NA
SPEM3        23.6370     -22.154765   3.02785 -7.3170030           NA           NA
AC022167.5   21.8451      -2.056240   2.89545 -0.7101618           NA           NA
BX276092.9   29.9636       0.407326   2.89048  0.1409199           NA           NA
ETDC         22.5675      -1.795274   2.89421 -0.6202983           NA           NA
> 
```

The ﬁrst three lines in this output show the contrast and the statistical test that were used to compute these results. Below these lines is the actual table with 6 columns:

+ **baseMean** represents the average normalized expression of the gene across all considered samples.
+ **log2FoldChange** represents the base-2 logarithm of the fold change of the normalized expression of the gene in the given contrast.
+ **lfcSE** represents the standard error of log2 fold change estimate.
+ **stat** is the statistic calculated in the contrast.
+ **pvalue** represent the pvalue.
+ **padj** represents the pvalue adjusted for multiple testing.


### Some more diagnostic plots

At this point, before going to further downstream analysis, it is important to look for the quality of our data in hand and to check if that has improved or not.

+ MA plot:

An MA plot is a useful tool to check if the data normalization worked well. The MA plot is a scatter plot where the x-axis denotes the average of normalized counts across samples and the y-axis denotes the log fold change in the given
contrast (here case vs control). Most points are expected to be on the horizontal 0 line (most genes are not expected to be differentially expressed).

```r
DESeq2::plotMA(object = dds, ylim = c(-5, 5))
```

![image5](https://user-images.githubusercontent.com/85447250/216890585-78fc1103-8854-46c5-b060-1575bf820b96.png)

Fig. MA plot of differential expression results.

+ p-value distribution

It is also important to observe the distribution of raw p-values (Figure 8.7). We expect to see a peak around low p-values and a uniform distribution at P-values above 0.1. Otherwise, adjustment for multiple testing would not work and the results
are not meaningful.

```r
ggplot(data = as.data.frame(DEresults), aes(x = pvalue)) +
  geom_histogram(bins = 100)
```

![image6](https://user-images.githubusercontent.com/85447250/216891706-db24f9b4-a9d8-4509-bcf7-77af4d14c1dc.png)

Fig. P-value distribution genes before adjusting for multiple testing

+ PCA plot

Another important diagnostic plot to look for the reproducibility of sample replicates is to plot the PCA for the data.

```r
rld <- rlog(dds)
DESeq2::plotPCA(rld, ntop = 500, intgroup = 'group') +
  ylim(-50, 50) + theme_bw()
```
![image7](https://user-images.githubusercontent.com/85447250/217051810-0bd43455-068c-473e-a1f5-5d6a73af1c21.png)

Fig. PCA plot of top 500 most variable genes.

+ Relative Log Expression plot:

This is a plot that check whether the data in hand is appropriatly normalized or not. Some time data does not get normalized even after applying aforementioned methods. This may be due to some unforseeable factor or variation, like the temparature of the day experiment was carried out, different persons carrying parts of experiments etc. It is quick diagnostic tools which can be applied to both the raw and normalized data to check if the data need further processing.

```r
library(EDASeq)
par(mfrow = c(1, 2))
plotRLE(countData, outline=FALSE, ylim=c(-4, 4),
        col=as.numeric(colData$group),
        main = 'Raw Counts')
plotRLE(DESeq2::counts(dds, normalized = TRUE),
        outline=FALSE, ylim=c(-4, 4),
        col = as.numeric(colData$group),
        main = 'Normalized Counts')
```
![image8](https://user-images.githubusercontent.com/85447250/217054806-5d71707e-306e-44bc-ae4a-418a7c84ccb5.png)

Fig. Relative log expression plots based on raw and normalized count matrices.

Here the RLE plot is comprised of boxplots, where each box-plot represents the distribution of the relative log expression of the genes expressed in the corresponding sample. Each gene’s expression is divided by the median expression value of that gene across all samples. Then this is transformed to log scale, which gives the relative log expression value for a single gene. The RLE values for all the genes from a sample are visualized as a boxplot. Ideally the boxplots are centered around the horizontal zero line and are as tightly distributed as possible. From the plots that we have made for the raw and normalized count data, we can observe how the normalized dataset has improved upon the raw count data for all the samples.
