In a typical Rna-seq experiment thousands of genes are found to be differentially expressed. Each gene may be involved in one or more biological processes, may be part of one or more pathways, or may share and/or part of more than one relevent biological phenomena. If we just go about documenting the effect caused by each gene individually to make the whole story, it will certainly be a cumborsome task if not impossible, and that too with a risk of concluding a contrasting/confusing results. To save us of this dilemma some smart people from [GO consortium](http://geneontology.org/)  came with a brilliant idea. The idea being that create a set of genes (a collection of genes) with a shared biological property, a comman pathways, and unsurprisingly a comman GO term etc. So instead of looking at the Diffrential Expression of individual genes between two groups (e.g. case and control) of samples, we look for the what GO Terms are enriched in our experiments. This gives us a broader view of the results at hand. Looking at the     
